src/graph/src/install.ts

```diff
@@ -30,30 +55,155 @@ export const install = async (
       throw err
     }
   }
-  const modifiers = GraphModifier.maybeLoad(options)
-
-  const act = actualLoad({
-    ...options,
-    mainManifest,
-    loadManifests: true,
-    modifiers: undefined, // modifiers should not be used here
-  })
-  const graph = await idealBuild({
-    ...options,
-    actual: act,
-    add,
-    mainManifest,
-    loadManifests: true,
-    modifiers,
-  })
-  const diff = await reify({
-    ...options,
-    add,
-    actual: act,
-    graph,
-    loadManifests: true,
-    modifiers,
-  })
-
-  return { graph, diff }
+
+  if (options.frozenLockfile) {
+    if (add?.modifiedDependencies) {
+      const dependencies: string[] = []
+      for (const [, deps] of add) {
+        for (const [name] of deps) {
+          dependencies.push(name)
+        }
+      }
+      throw error(
+        'Cannot add dependencies when using --frozen-lockfile',
+        { found: dependencies.join(', ') },
+      )
+    }
+
+    const lockfileGraph = loadVirtual({
+      ...options,
+      mainManifest,
+      skipLoadingNodesOnModifiersChange: false,
+    })
+
+    const emptyAdd = Object.assign(
+      new Map<DepID, Map<string, Dependency>>(),
+      { modifiedDependencies: false },
+    )
+    const emptyRemove = Object.assign(new Map<DepID, Set<string>>(), {
+      modifiedDependencies: false,
+    })
+    const importerSpecs = getImporterSpecs({
+      graph: lockfileGraph,
+      add: emptyAdd,
+      remove: emptyRemove,
+      ...options,
+    })
+
+    // Check for spec changes by comparing package.json specs with lockfile edges
+    const specChanges: string[] = []
+    for (const importer of lockfileGraph.importers) {
+      const deps = getDependencies(importer, options)
+      for (const [depName, dep] of deps) {
+        const edge = importer.edgesOut.get(depName)
+        if (edge?.spec) {
+          const manifestSpec = Spec.parse(
+            depName,
+            dep.spec.toString(),
+            options,
+          )
```

Comment:
one very minor clean up, we can reduce the boilerplate even more here (and avoid the extra `Spec.parse` call) as `dep` is an object with a `spec` and `type` properties. `spec` is the parsed spec directly from the manifest, so this could be further trimmed down to something like:

```suggestion
        const edge = importer.edgesOut.get(depName)
        if (edge?.spec) {
          const manifestSpec = dep.spec
```